File: src/coddy/tools/definitions.py
"""Tool definitions for LLM tool calling.

These definitions follow the OpenAI function calling format.
"""

from typing import Any


def get_all_tools() -> list[dict[str, Any]]:
    """Get all available tool definitions.
    
    Returns:
        List of tool definitions in OpenAI format.
    """
    return [
        EXECUTE_CODE_TOOL,
        READ_FILE_TOOL,
        WRITE_FILE_TOOL,
        LIST_FILES_TOOL,
    ]


# Primary tool: Execute code in the sandbox
EXECUTE_CODE_TOOL: dict[str, Any] = {
    "type": "function",
    "function": {
        "name": "execute_code",
        "description": (
            "Execute code in a sandboxed environment. Use this for calculations, "
            "data processing, file creation, plotting, or any task that benefits "
            "from running code. The environment is stateful: variables and files "
            "persist between calls within the same session."
        ),
        "parameters": {
            "type": "object",
            "properties": {
                "language": {
                    "type": "string",
                    "enum": ["python", "nodejs"],
                    "description": (
                        "The programming language to use. Use 'python' for Python 3.12 "
                        "or 'nodejs' for Node.js 20."
                    ),
                },
                "code": {
                    "type": "string",
                    "description": (
                        "The code to execute. For Python, use print() to show output. "
                        "For Node.js, use console.log(). Code should be complete and "
                        "self-contained."
                    ),
                },
            },
            "required": ["language", "code"],
        },
    },
}

# Read a file from the sandbox
READ_FILE_TOOL: dict[str, Any] = {
    "type": "function",
    "function": {
        "name": "read_file",
        "description": "Read the contents of a file from the sandbox filesystem.",
        "parameters": {
            "type": "object",
            "properties": {
                "path": {
                    "type": "string",
                    "description": (
                        "The path to the file to read, relative to /home/user or "
                        "absolute. Examples: 'data.csv', '/home/user/output.txt'"
                    ),
                },
            },
            "required": ["path"],
        },
    },
}

# Write a file to the sandbox
WRITE_FILE_TOOL: dict[str, Any] = {
    "type": "function",
    "function": {
        "name": "write_file",
        "description": (
            "Write content to a file in the sandbox filesystem. "
            "Creates the file if it doesn't exist, overwrites if it does."
        ),
        "parameters": {
            "type": "object",
            "properties": {
                "path": {
                    "type": "string",
                    "description": (
                        "The path to write the file to, relative to /home/user or "
                        "absolute. Examples: 'output.txt', '/home/user/data.json'"
                    ),
                },
                "content": {
                    "type": "string",
                    "description": "The content to write to the file.",
                },
            },
            "required": ["path", "content"],
        },
    },
}

# List files in the sandbox
LIST_FILES_TOOL: dict[str, Any] = {
    "type": "function",
    "function": {
        "name": "list_files",
        "description": "List files in a directory in the sandbox filesystem.",
        "parameters": {
            "type": "object",
            "properties": {
                "path": {
                    "type": "string",
                    "description": (
                        "The directory path to list, relative to /home/user or "
                        "absolute. Defaults to /home/user if not specified."
                    ),
                },
            },
        },
    },
}

File: internal/tools/definitions.go
// Package tools defines LLM tool schemas.
package tools

import (
	"encoding/json"

	"github.com/firasmosbehi/coddy/pkg/models"
)

// GetDefinitions returns all available tool definitions.
func GetDefinitions() []models.ToolDefinition {
	return []models.ToolDefinition{
		ExecuteCodeTool(),
		ReadFileTool(),
		WriteFileTool(),
		ListFilesTool(),
	}
}

// ExecuteCodeTool returns the execute_code tool definition.
func ExecuteCodeTool() models.ToolDefinition {
	params, _ := json.Marshal(map[string]interface{}{
		"type": "object",
		"properties": map[string]interface{}{
			"language": map[string]interface{}{
				"type":        "string",
				"enum":        []string{"python", "nodejs"},
				"description": "The programming language to use",
			},
			"code": map[string]interface{}{
				"type":        "string",
				"description": "The code to execute. Use print() in Python or console.log() in Node.js to show output.",
			},
		},
		"required": []string{"language", "code"},
	})

	return models.ToolDefinition{
		Type: "function",
		Function: models.FunctionDef{
			Name:        "execute_code",
			Description: "Execute code in a sandboxed environment. Use this for calculations, data processing, file creation, plotting, or any task that benefits from running code.",
			Parameters:  params,
		},
	}
}

// ReadFileTool returns the read_file tool definition.
func ReadFileTool() models.ToolDefinition {
	params, _ := json.Marshal(map[string]interface{}{
		"type": "object",
		"properties": map[string]interface{}{
			"path": map[string]interface{}{
				"type":        "string",
				"description": "The path to the file to read",
			},
		},
		"required": []string{"path"},
	})

	return models.ToolDefinition{
		Type: "function",
		Function: models.FunctionDef{
			Name:        "read_file",
			Description: "Read the contents of a file from the sandbox filesystem.",
			Parameters:  params,
		},
	}
}

// WriteFileTool returns the write_file tool definition.
func WriteFileTool() models.ToolDefinition {
	params, _ := json.Marshal(map[string]interface{}{
		"type": "object",
		"properties": map[string]interface{}{
			"path": map[string]interface{}{
				"type":        "string",
				"description": "The path to write the file to",
			},
			"content": map[string]interface{}{
				"type":        "string",
				"description": "The content to write to the file",
			},
		},
		"required": []string{"path", "content"},
	})

	return models.ToolDefinition{
		Type: "function",
		Function: models.FunctionDef{
			Name:        "write_file",
			Description: "Write content to a file in the sandbox filesystem.",
			Parameters:  params,
		},
	}
}

// ListFilesTool returns the list_files tool definition.
func ListFilesTool() models.ToolDefinition {
	params, _ := json.Marshal(map[string]interface{}{
		"type": "object",
		"properties": map[string]interface{}{
			"path": map[string]interface{}{
				"type":        "string",
				"description": "The directory path to list (defaults to /home/user)",
			},
		},
	})

	return models.ToolDefinition{
		Type: "function",
		Function: models.FunctionDef{
			Name:        "list_files",
			Description: "List files in a directory in the sandbox filesystem.",
			Parameters:  params,
		},
	}
}

// GetToolNames returns a list of available tool names.
func GetToolNames() []string {
	defs := GetDefinitions()
	names := make([]string, len(defs))
	for i, def := range defs {
		names[i] = def.Function.Name
	}
	return names
}

// FindTool finds a tool by name.
func FindTool(name string) (models.ToolDefinition, bool) {
	for _, def := range GetDefinitions() {
		if def.Function.Name == name {
			return def, true
		}
	}
	return models.ToolDefinition{}, false
}
